test/regtest.py

```diff
@@ -69,78 +79,171 @@ def metadata(file):
+    logging.warning("@expect MISSING IN %s" % file)
     m['expect'] = 'verified'
 
   return m
-  
-parser = argparse.ArgumentParser()
-parser.add_argument("--exhaustive", help="be exhaustive", action="store_true")
-args = parser.parse_args()
-
-print "Running regression tests..."
-print
-
-passed = failed = timeouts = unknowns = 0
-
-try:
-  for test in glob.glob("./**/*.c") if args.exhaustive else glob.glob("./basic/*.c"):
-    meta = metadata(test)
-
-    if meta['skip']:
-      continue
-
-    print "{0:>20}".format(test)
-    sys.stdout.flush()
-
-    cmd = ['smackverify.py', test]
-    cmd += ['--time-limit', str(meta['time-limit'])]
-    cmd += ['-o', test + '.bpl']
-    cmd += meta['flags']
-
-    for memory in meta['memory'][:100 if args.exhaustive else 1]:
-      cmd += ['--mem-mod=' + memory]
-
-      for verifier in meta['verifiers'][:100 if args.exhaustive else 1]:
-        cmd += ['--verifier=' + verifier]
-
-        print "{0:>20} {1:>10}    :".format(memory, verifier),
-
-        t0 = time.time()
-        p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-        out, err  = p.communicate()
-        elapsed = time.time() - t0
-
-        result = get_result(out+err)
-
-        if result == meta['expect']:
-          print green('PASSED '),
-          passed += 1
-
-        elif result == 'timeout':
-          print red('TIMEOUT'),
-          timeouts += 1
-
-        elif result == 'unknown':
-          print red('UNKNOWN'),
-          unknowns += 1
-
-        else:
-          print red('FAILED '),
-          failed += 1
-
-        print '  [%.2fs]' % round(elapsed, 2)
-        sys.stdout.flush()
-  
-except KeyboardInterrupt:
-  pass
-
-print
-print ' PASSED count:', passed
-print ' FAILED count:', failed
-
-if timeouts > 0:
-  print 'TIMEOUT count:', timeouts
 
-if unknowns > 0:
-  print 'UNKNOWN count:', unknowns
+def process_test(cmd, test, memory, verifier, meta):
+    """
+    This is the worker function for each process. This function process the supplied
+    test and returns an integer indicating the test results.
+
+    :return: 0 (PASSED), -1 (FAILED), 1 (TIMEDOUT), 2 (UNKNOWN)
+    """
+    str_result = "{0:>20}\n".format(test)
+    str_result += "{0:>20} {1:>10}    :".format(memory, verifier)
+
+    t0 = time.time()
+    p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    out, err  = p.communicate()
+    elapsed = time.time() - t0
+
+    # get the test results
+    result = get_result(out+err)
+    passed = timedout = unknown = False
+    if result == meta['expect']:
+      str_result += 'PASSED '
+      passed = True
+
+    elif result == 'timeout':
+      str_result += 'TIMEOUT'
+      timedout = True
+
+    elif result == 'unknown':
+      str_result += 'UNKNOWN'
+      unknown = True
+
+    else:
+      str_result += 'FAILED '
+
+    str_result += '  [%.2fs]' % round(elapsed, 2)
+
+    # log the info for this test
+    logging.info(str_result)
```

Comment:
Better to do logging in the `tally_result` callback, making it totally clear that the `logging.info(…)` calls are executed serially.